src/public/modules/forms/helpers/decryption.worker.js

```diff
@@ -84,13 +85,51 @@ function verifySignature(decryptedSubmission, created) {
   return verified.every((v) => v)
 }
 
-/** @typedef {{
- *    created: string,
- *    id: string,
- *    status?: string
- *    submissionData?: Object
- * }}
- * CsvRecord */
+class CsvRecord {
+  constructor(id, created, status) {
+    this.id = id
+    this.created = created
+    this.status = status
+
+    this._statusMessage = status
+    this._record = []
+
+    this.updateSubmissionData()
+  }
+
+  setStatus(status, msg) {
+    this.status = status
+    this._statusMessage = msg
+    this.updateSubmissionData()
+  }
+
+  setDownloadBlob(blob) {
+    this.downloadBlob = blob
+  }
+
+  setRecord(record) {
+    this._record = record
+    this.updateSubmissionData()
+  }
+
+  updateSubmissionData() {
+    let downloadStatus = {
+      _id: '000000000000000000000000',
+      fieldType: 'textfield',
+      question: 'Download Status',
+      answer: this._statusMessage,
+      questionNumber: 1000000,
+    }
+    let output = cloneDeep(this._record)
+    output.unshift(downloadStatus)
+
+    this.submissionData = {
+      created: this.created,
+      submissionId: this.id,
+      record: output,
+    }
```

Comment:
would it be better to convert this function into a getter method, so that there is no need to hold derived state in `submissionData`?

Reply:
So the `postMessage` call gets rid of all JS methods in the object to be passed back (you only end up with the simple properties), so we have to materialize the `submissionData` field at some point within the worker before `postMessage`.

We can materialize it right before the `postMessage` call though (i.e. we can manually call `materialize()` right before `postMessage` instead of calling it automatically)

Reply:
I see, thanks for explaining! let's go with this - could you help to JSDoc this behavior too?

Reply:
you could also pass strings between the main thread and web workers, and use `JSON.parse` and `JSON.stringify` to serialise/deserialise

Reply:
Added